### DISCUSSION TITLE
URL change when using SSG with query parameters

### DISCUSSION THREAD

**Discussion by nikolajajac** [vercel/next.js#47389] [Help]
### Summary

We have a NextJS application that acts as a renderer for content from a headless CMS. The application should appear as an integral part of a larger website. The website has a load balancer in front of it, which we use to forward certain paths to the application. The application is configured with a base path (e.g. `/content`), and then we have a couple of forwarding rules that forward for example `www.example.com/test-page` to `www.example.com/content/test-page` and then to the NextJS application, so that the `/content` base path is not visible to the end users.

The pages are rendered with SSG, and everything works fine until a query parameter is added when requesting a page. The page is returned and displays fine, but then it seems that some NextJS client code runs that changes the URL in the browser address bar, which is unexpected and unwanted in our case. So for example, if one requests `www.example.com/test-page?param=123`, the page is retrieved, and just before it ends to load, the URL changes to `www.example.com/content/test-page?param=123`. As if the router "reinitializes", and changes the URL to what it thinks it should be.

Does anyone know why this happens, and how can we maybe prevent it?

**Comment by icyJoseph** [comment]
Next.js version?

I wonder if this is because of: https://nextjs.org/docs/advanced-features/automatic-static-optimization#how-it-works

> During prerendering, the router's query object will be empty since we do not have query information to provide during this phase. After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

- How have you configured the base path?

> we have a couple of forwarding rules that forward for example www.example.com/test-page to www.example.com/content/test-page and then to the NextJS application

So Next.js has no rewrite/redirects configured itself?

**Reply by nikolajajac** [reply]
Next.js version is 13.2.3. The base path is configured in `next.config.js`, pretty standard. No, the forwarding I mentioned all happens in the load balancer. And yes, I assume it is somehow related to the process you linked...

**Reply by icyJoseph** [reply]
It feels like in one hand, we tell Next.js that it renders at `/content`, but in the other, we render it at `/`.

If I understand correctly this is only for a couple of pages, like `/test-page`, should be rendered outside the `basePath` scope, and you use a load balancer to make that swap. 

I feel like I don't have enough info, but https://nextjs.org/docs/api-reference/next.config.js/rewrites#rewrites-with-basepath-support, this documents a way to rewrite, with basePath enable/disabled, so maybe that's what you need to use? 

I understand that the load balancer is probably there to separate concerns, and let the rendering server render, while others take care of routing the requests, but perhaps this requires an exception to that rule?

**Reply by nikolajajac** [reply]
Thanks for your help @icyJoseph! It's not single pages, we actually forward whole groups with all the sub-paths. E.g. the forwarding rules are as follows `www.example.com/group-1/*` -> `www.example.com/content/group-1/*` -> `internal-address/content/group-1/*`. And under `group-1` there can be many pages, e.g. `group-1/page-1`, `group-1/sub-group-2/page-2`, etc. Rewriting all those paths does not seem right.

**Reply by nikolajajac** [reply]
The solution we ended up with was to overwrite the `changeState` method of the router provided in the props of the custom app (`_app.tsx`).  In the method, we change the URL that will be set in the browser so that it does not contain the base path, if this is not desired.

**Reply by kkomitski** [reply]
Has there been any cleaner solution to this since the last comment? Overwriting the `changeState` seems a bit clunky?
